### PR TITLE
Fixing Navigation end event loss of cancelling

### PR DESCRIPTION
**Pull request by Udumft**
vk-694-cancel-event-loss: added forcing EventsManager datasource to o……verride nilling weak reference during deallocation.

### Description
Resolves an issue, where `Navigation end` event was not logged if trips was cancelled.

### Implementation
Issue is caused by nilling weak reference to a `NavigationService` during it's deinit stage. To avoid that I've added forcing a strong reference during event logging.

### CODE REVIEW

File: Sources/MapboxCoreNavigation/NavigationService.swift [hunk 1]
```diff
@@ -266,7 +266,9 @@ public class MapboxNavigationService: NSObject, NavigationService {
     
     deinit {
         suspendNotifications()
-        endNavigation()
+        eventsManager.withBackupDataSource(self) {
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/NavigationService.swift:269]: Looks too hacky, but I can't really come to a better solution. Making `dataSource` strong seems to be even worse.

File: Sources/MapboxCoreNavigation/NavigationEventsManager.swift [hunk 2]
```diff
@@ -22,7 +22,23 @@ open class NavigationEventsManager {
     
     var outstandingFeedbackEvents = [CoreFeedbackEvent]()
     
-    weak var dataSource: EventsManagerDataSource?
+    func withBackupDataSource(_ forcedDataSource: EventsManagerDataSource, action: () -> Swift.Void) {
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/NavigationEventsManager.swift:25]:
```suggestion
    func withBackupDataSource(_ forcedDataSource: EventsManagerDataSource, action: () -> Void) {
```

File: Sources/MapboxCoreNavigation/NavigationService.swift [hunk 3]
```diff
@@ -266,7 +266,9 @@ public class MapboxNavigationService: NSObject, NavigationService {
...
+            endNavigation()
```
Comment (bamx23) [inline Sources/MapboxCoreNavigation/NavigationService.swift:270]: NIT: A potential issue might be if there are async operations inside

**Review by bamx23 — APPROVED** [review]

File: Sources/MapboxCoreNavigation/NavigationService.swift (hunk 3 above)
Comment (Udumft) [reply Sources/MapboxCoreNavigation/NavigationService.swift:270]: True. But I don't see any better options

File: Sources/MapboxCoreNavigation/NavigationService.swift (hunk 1 above)
Comment (Udumft) [reply Sources/MapboxCoreNavigation/NavigationService.swift:269]: Same story...

**Pull request merged**